girder/models/file.py

```diff
@@ -93,23 +94,62 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type contentDisposition: str or None
         :type extraParameters: str or None
         """
+        if offset == 0:
+            events.trigger('model.file.download.before', info={'file': file})
+        else:
+            events.trigger('model.file.download.during', info={'file': file})
+
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
-                file, offset=offset, headers=headers, endByte=endByte,
-                contentDisposition=contentDisposition,
-                extraParameters=extraParameters)
+            try:
+                return self._handleAssetstoreDownload(file, offset=offset, headers=headers,
+                                                      endByte=endByte,
+                                                      contentDisposition=contentDisposition,
+                                                      extraParameters=extraParameters)
+            except cherrypy.HTTPRedirect:
+                events.trigger('model.file.download.after', info={'file': file, 'redirect': True})
+                raise
         elif file.get('linkUrl'):
             if headers:
+                events.trigger('model.file.download.after', info={'file': file, 'redirect': True})
                 raise cherrypy.HTTPRedirect(file['linkUrl'])
             else:
                 endByte = endByte or len(file['linkUrl'])
 
                 def stream():
                     yield file['linkUrl'][offset:endByte]
+                    if endByte >= len(file['linkUrl']):
+                        events.trigger('model.file.download.after',
+                                       info={'file': file, 'redirect': False})
                 return stream
         else:  # pragma: no cover
             raise Exception('File has no known download mechanism.')
 
+    def _handleAssetstoreDownload(self, file, offset=0, headers=True, endByte=None,
+                                  contentDisposition=None, extraParameters=None):
+        fileDownload = self.getAssetstoreAdapter(file).downloadFile(
+            file, offset=offset, headers=headers, endByte=endByte,
+            contentDisposition=contentDisposition,
+            extraParameters=extraParameters)
+        if inspect.isgeneratorfunction(fileDownload) or inspect.isgenerator(fileDownload):
+            def downloadGenerator():
+                # Get generator object from function if necessary
+                if inspect.isgeneratorfunction(fileDownload):
```

Comment:
I don't think this test is necessary, won't it work in either case if you use `for data in gen()`?